File: gpuclock.py
import subprocess
import os

class GPUClock:
	def __init__(self):
		self.path = "/sys/devices/system/cpu"
		self.gpu_freq = 0
		self.gpu_util = 0
	
	def get_clock(self):
		# print("... Getting the GPU frequency")
		path = os.path.join(os.getcwd() + "/scripts/gpu_get_clock.sh")
		cmd = path 
		gpu_freq = subprocess.check_output(cmd,shell=True)
		gpu_freq = gpu_freq.decode('ascii')
		# gpu_freq = int(gpu_freq)
		# print("GPUFreq is : "+str(gpu_freq))
		# print(f"The GPU frequency is {gpu_freq}")
		if len(gpu_freq) < 2:
			return -1
		self.gpu_freq = int(gpu_freq[:-1])
		return self.gpu_freq

	def adb_get_clock(self):
		path = os.path.join(os.getcwd() + "/scripts/adb_get_gpufreq.sh")
		cmd = path 
		gpu_freq = subprocess.check_output(cmd,shell=True)
		gpu_freq = gpu_freq.decode('ascii')
		# gpu_freq = int(gpu_freq)
		# print("GPUFreq is : "+str(gpu_freq))
		# print(f"The GPU frequency is {gpu_freq}")
		if len(gpu_freq) < 2:
			return -1
		self.gpu_freq = int(gpu_freq[:-1])
		return self.gpu_freq		
	
	def set_clock(self, freq):
		# print("... Setting the GPU to"+str(freq))
		if freq != self.gpu_freq:
			path = os.path.join(os.getcwd() + "/scripts/gpu_set_clock.sh")
			inc_flag = 0
			if freq > self.gpu_freq:
				inc_flag = 1
			cmd = path+" "+str(freq)+" "+str(inc_flag)
			subprocess.check_call(cmd, shell=True)
			self.gpu_freq = freq
			# print("The clock is in GPU to"+str(freq))		
		return
	
	def adb_set_clock(self, freq):
		if freq != self.gpu_freq:
			path = os.path.join(os.getcwd() + "/scripts/adb_set_gpu.sh")
			inc_flag = 0
			if freq > self.gpu_freq:
				inc_flag = 1
			cmd = path+" "+str(freq)+" "+str(inc_flag)
			subprocess.check_call(cmd, shell=True)
			self.gpu_freq = freq
			# print("The clock is in GPU to"+str(freq))		
		return

	
	def get_all_clock(self):
		print("... Getting the GPU all frequency")
		# path = os.path.join(os.getcwd() + "/scripts/gpu_get_all_clock.sh")
		# cmd = path
		# freq_list = subprocess.check_output(cmd,shell=True)
		# freq_list = freq_list.split(" ")
		# print(f"The frequency list of {freq_list}")
		# self.cpu_freq = curr_freq
		# if len(freq_list) <= 0:
		#     return []
		# return freq_list
		return []

	def get_utilization(self):
		print("... Getting the utilization of the gpu")
		# path = os.path.join(os.getcwd() + "/scripts/gpu_get_util.sh")
		# cmd = path
		# gpu_util  = subprocess.check_output(cmd,shell=True)
		gpu_util = -1
		return gpu_util

File: code_translated_to_cpp/gpuclock.cpp
#include "gpuclock.h"

GpuClock::GpuClock() {
    this->path_     = "/sys/devices/system/cpu";
    this->gpu_freq_ = 0;
    this->gpu_util_ = 0.0;
}

long long GpuClock::GetClock() {
    std::cout << "Getting the GPU frequency" << std::endl;

    char command[256];
    // ISO C++ forbids converting a string constant to ‘char*’, cannot set char* command = "..."
    sprintf(command, "bash ../scripts/gpu_get_clock.sh");

    char buffer[128];

    // Based on https://stackoverflow.com/a/478960
    FILE* pipe = popen(command, "r");
    if (pipe == NULL) {
        std::cout << "popen() failed while trying to get the GPU frequency" << std::endl;
        exit(1);
    }
    else {
        if (fgets(buffer, sizeof(buffer), pipe) != NULL) {
            // There was some error checking here in Python code?????
            this->gpu_freq_ = atoll(buffer); // Newline at end is fine???? Format of files being read???
            // std::cout << "The GPU frequency is " << this->gpu_freq_ << std::endl;
        }
        else {
            std::cout << "fgets() failed while trying to get the GPU frequency" << std::endl; 
            exit(1);
        }
    }
    pclose(pipe);

    return this->gpu_freq_;
}

void GpuClock::SetClock(long long new_freq) {
    // std::cout << "Setting GPU frequency to " << new_freq << std::endl;

    if (new_freq != this->gpu_freq_) {
        int increase_flag = 0;
        if (new_freq > this->gpu_freq_) {
            increase_flag = 1;
        }

        char command[256];
        sprintf(command, "bash ../scripts/gpu_set_clock.sh %lld %d", new_freq, increase_flag);

        if (system(command) != 0) {
            std::cout << "Could not set GPU clock frequency" << std::endl;
            exit(1);
        }

        this->gpu_freq_ = new_freq;
    }

    return;
}

double GpuClock::GetUtilization() {
    return -1.0;
}